### ISSUE TITLE
Effective judgement of whether quick alignment or not for this project by the plot

### ISSUE THREAD

**Issue by begeekmyfriend** [Rayhane-mamah/Tacotron-2#342]
Hopeless for quick or even achieving alignment. It seems that the gradient is not towards the direction of the optimal one.
![step-1000-align](https://user-images.githubusercontent.com/6031938/53691695-57f0f300-3dbe-11e9-986c-3fbcdfab8b64.png)
Sign for quick alignment. Note the location of the shiny pixels.
![step-1000-align](https://user-images.githubusercontent.com/6031938/53691679-e4e77c80-3dbd-11e9-8a65-e59b340f1b28.png)
Strong sign for quick alignment.
![step-1000-align](https://user-images.githubusercontent.com/6031938/53691728-f9784480-3dbe-11e9-9159-025b40ddb100.png)
Above all have been testified by different dataset. See https://github.com/Rayhane-mamah/Tacotron-2/issues/336

**Comment by alexdemartos** [comment]
Hey @begeekmyfriend ,

something related to this. Are you aware of this problem?

ref: https://github.com/mozilla/TTS/issues/50

Do you think this issue might be also affecting alignments in this repo?

PS: I somehow feel that attention in this model is not very strong for datasets more difficult than LJSpeech. Any ideas on how to improve it in addition to switch to L1 loss? Maybe making the PreNet bottleneck bigger/smaller? I've heard of different alignment models (like GMM improving stability) but somehow decreasing speech quality :\

PS2: I am having some thoughts. During training, the loss doesn't really care if the attention is learnt or not. The decoder can learn to generate the next output mels by just feeding the previous mels (teacher forcing), so the loss keeps improving even if the attention model has no clue. What if we could add a tunable loss penalty whenever the alignment is not really diagonal/monotonic? This penalty weight can decay in the same manner learning rate decays, so we don't force the model to learn monotonic alignments (just at the beggining to guide it)

**Comment by begeekmyfriend** [comment]
Loss function just cares the differences between values of ground truth and prediction while alignment dipicts the measures of similarity between the mel spectrograms and texts. And therefore the loss function does nothing with the alignment in my opinion. Besides there is no obvious evidence currenly that indicates the impacts on quality of the synthesized speech in different kinds of attention. And the loss penity on the measures of similarity between the output of the encoder and the decoder is hard to design.

**Comment by begeekmyfriend** [comment]
Prenet is helpful to the representation of the texts at the encoder and the measures of similarity between the texts and the mel spectrograms at the decoder. So we cannot directly feed the embeddings into both the encoder and the decoder because that will impair the alignment learning.

**Comment by alexdemartos** [comment]
Just found that this alignment-loss-penalty idea is not new:

https://github.com/mozilla/TTS/issues/13#issuecomment-384276551

**Comment by begeekmyfriend** [comment]
I know. That is guided attention from [DC-TTS](https://github.com/begeekmyfriend/dc_tts/blob/master/train.py#L91-L95). I have tried that project but the evaluation is not good enough compared to this one.

**Comment by zhangyi02** [comment]
I adopt your guided attention loss, but still can't get any alignment. As the images shows, the alignments  is just a line that seems never to convergence. Have you ever met this problem, or any suggestions?
By the way, I'm using L1 loss.
![image](https://user-images.githubusercontent.com/41054719/55711161-b85afc00-5a1e-11e9-8815-662d9f30e078.png)
![image](https://user-images.githubusercontent.com/41054719/55711197-ce68bc80-5a1e-11e9-8d43-4c14c2008028.png)

**Comment by begeekmyfriend** [comment]
@zhangyi02 Please adapt `max_text_length` and `max_frame_num` to the best ratio of N:T according to your dataset. You might give up some exceptional wave clips.

**Comment by zhangyi02** [comment]
@begeekmyfriend 
I just adopt your code, since I am using Chinese mandarin too, or how could I decide which ratio to use?
And what's your mean about “exceptional wave clips”? Long mels and texts? 
```
N = self._hparams.max_text_length
T = self._hparams.max_mel_frames // self._hparams.outputs_per_step
```

**Comment by begeekmyfriend** [comment]
Typically the `max_text_length` and `max_frame_num` is just the max length of your dataset. I would round up these value to hundred. This [line](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-guided-attention/datasets/preprocessor.py#L89-L90) would drop some wav clips that exceed the hyper parameters. Please check out the latest commit and try again.
